File: shoring/compat/src/test/java/com/panopset/compat/test/RandomopTest.java
package com.panopset.compat.test;

import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.panopset.compat.Randomop;

public class RandomopTest {

  @Test
  void test() {
    int found = 0;
    while (found != 7) {
      int result = Randomop.random(1, 3);
      found = verifyInRange1to3(result, found);
    }
  }

  @Test
  void testMinGreaterThanMax() {
    int found = 0;
    while (found != 7) {
      int result = Randomop.random(3, 1);
      found = verifyInRange1to3(result, found);
    }
  }

  @Test
  void testSingle() {
    int result = Randomop.random(0, 0);
    Assertions.assertEquals(Integer.valueOf(0), result);
  }

  @Test
  void testBytes() {
    byte[] ba = new byte[2];
    Randomop.nextBytes(ba);
    Assertions.assertNotNull(ba);
  }

  @Test
  void testNextInt() {
    int x = Randomop.nextInt();
    int lastOne = x;
    x = Randomop.nextInt();
    Assertions.assertNotEquals(lastOne, x);
  }

  @Test
  void testNextLong() {
    long x = Randomop.nextLong();
    long lastOne = x;
    x = Randomop.nextLong();
    Assertions.assertNotEquals(lastOne, x);
  }

  private static int verifyInRange1to3(int result, int found) {
    int rtn = found;
    MatcherAssert.assertThat(4, Matchers.greaterThan(result));
    MatcherAssert.assertThat(0, Matchers.lessThan(result));
    switch (result) {
      case 1:
        rtn |= 1;
        break;
      case 2:
        rtn |= 2;
        break;
      case 3:
        rtn |= 4;
        break;
    }
    return rtn;
  }
}

File: code/shoring/compat/src/test/kotlin/com/panopset/compat/test/RandomopTest.kt
package com.panopset.compat.test

import com.panopset.compat.Randomop
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class RandomopTest {
    @Test
    fun test() {
        var found = 0
        while (found != 7) {
            val result = Randomop.random(1, 3)
            found = verifyInRange1to3(result, found)
        }
    }

    @Test
    fun testMinGreaterThanMax() {
        var found = 0
        while (found != 7) {
            val result = Randomop.random(3, 1)
            found = verifyInRange1to3(result, found)
        }
    }

    @Test
    fun testSingle() {
        val result = Randomop.random(0, 0)
        Assertions.assertEquals(Integer.valueOf(0), result)
    }

    @Test
    fun testBytes() {
        val ba = ByteArray(2)
        Randomop.nextBytes(ba)
        Assertions.assertNotNull(ba)
    }

    @Test
    fun testNextInt() {
        var x = Randomop.nextInt()
        val lastOne = x
        x = Randomop.nextInt()
        Assertions.assertNotEquals(lastOne, x)
    }

    @Test
    fun testNextLong() {
        var x = Randomop.nextLong()
        val lastOne = x
        x = Randomop.nextLong()
        Assertions.assertNotEquals(lastOne, x)
    }

    companion object {
        private fun verifyInRange1to3(result: Int, found: Int): Int {
            var rtn = found
            Assertions.assertTrue(4 > result)
            Assertions.assertTrue(0 < result)
            when (result) {
                1 -> rtn = rtn or 1
                2 -> rtn = rtn or 2
                3 -> rtn = rtn or 4
            }
            return rtn
        }
    }
}
